### PR TITLE
Add Run As Dataflow Pipeline UI Components

### PR DESCRIPTION
**Pull request by tgroh**

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/DataflowPipelineLaunchConfigurationTabGroup.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+/**
+ * The Launch Configuration Tab Group for Dataflow Pipelines.
+ *
+ * <p>Provides Run Locally and Run Remotely configuration for launching a Dataflow Pipeline.
+ */
+public class DataflowPipelineLaunchConfigurationTabGroup extends
+    AbstractLaunchConfigurationTabGroup {
+
+  @Override
+  public void createTabs(ILaunchConfigurationDialog dialog, String mode) {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/DataflowPipelineLaunchConfigurationTabGroup.java]: This class can be tested. See AppEngineTabGroupTest for an example.

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/DefaultedPipelineOptionsComponent.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+import org.eclipse.swt.layout.GridData;
+import org.eclipse.swt.layout.GridLayout;
+import org.eclipse.swt.widgets.Button;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Group;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A Group that contains pipeline options that are can be defaulted with {@link
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/DefaultedPipelineOptionsComponent.java]: This is not a group though. I.e. it does not extend Group.

also "are can be" -> can be

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/DefaultedPipelineOptionsComponent.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+  private RunOptionsDefaultsComponent defaultOptions;
+
+  public DefaultedPipelineOptionsComponent(Composite parent, Object layoutData,
+      MessageTarget messageTarget, DataflowPreferences preferences) {
+    this.preferences = preferences;
+    customValues = new HashMap<>();
+
+    defaultsGroup = new Group(parent, SWT.NULL);
+    final int numColumns = 3;
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/DefaultedPipelineOptionsComponent.java]: doesn't need to be final

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/DefaultedPipelineOptionsComponent.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+import org.eclipse.swt.widgets.Group;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A Group that contains pipeline options that are can be defaulted with {@link
+ * DataflowPreferences}. Contains a Button (with Checkbox style) to enable and disable use of
+ * defaults and a {@link RunOptionsDefaultsComponent} to input said defaults.
+ */
+public class DefaultedPipelineOptionsComponent {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/DefaultedPipelineOptionsComponent.java]: This should be testable. See StandardDeployPreferencesPanelTest for an example of using ShellTestResouce

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/LaunchPipelineShortcut.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+    public LaunchableResource(IResource resource) {
+      this(resource, null, null);
+    }
+
+    public LaunchableResource(IResource resource, IMethod mainMethod, IType primaryType) {
+      this.resource = resource;
+      if (mainMethod == null && primaryType != null) {
+        throw new NullPointerException(
+            "mainMethod and primaryType must either both be null or neither can be null");
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/LaunchPipelineShortcut.java]: There should be two different exception messages here and below.This one could be

mainMethod is null but primaryType is not null

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/LaunchPipelineShortcut.java [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+  private ILaunchManager getLaunchManager() {
+    return DebugPlugin.getDefault().getLaunchManager();
+  }
+
+  /**
+   * A resource this shortcut is capable of launching. If the resource is a Java class file that
+   * contains a main method, contains a handle to that method and the primary type.
+   */
+  private static class LaunchableResource {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/LaunchPipelineShortcut.java]: Consider pulling out to a non-public outer class and adding tests.

File: plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineOptionsFormComponent.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+  private final Set<String> filterProperties;
+
+  private final Composite parent;
+  private FormToolkit formToolkit;
+  private Form form;
+  private Map<ExpandableComposite, LabeledTextMapComponent> optionsComponents;
+
+  public PipelineOptionsFormComponent(
+      Composite parent, String argumentSeparator, Set<String> filterProperties) {
+    this.modifyListeners = new HashSet<>();
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineOptionsFormComponent.java]: perhaps initialize the fields directly instead

**Comment by tgroh** [comment]
That should be fixed; I failed to push the appropriate updates.

**Pull request merged**